File: js/src/ec.js
/**
 * admitiq/ec — optional asymmetric (ES256 / ECDSA P-256) signing support.
 *
 * Use this instead of core.issue/verify when the verifier (e.g. a scanner
 * app on someone's phone, or a public endpoint) should NOT hold the same
 * secret used to issue tokens. The issuer signs with a private key; anyone
 * holding only the public key can verify a token but cannot forge new ones.
 *
 * Uses Node's built-in 'crypto' module — no extra dependency required.
 */
const crypto = require("crypto");
const {
  InvalidSignatureError,
  TokenExpiredError,
  TokenRevokedError,
  assertSupportedVersion,
} = require("./core");

function b64urlEncode(buffer) {
  return buffer
    .toString("base64")
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=+$/, "");
}

function b64urlDecode(str) {
  let padded = str.replace(/-/g, "+").replace(/_/g, "/");
  while (padded.length % 4) padded += "=";
  return Buffer.from(padded, "base64");
}

/**
 * Generate a new EC (P-256) keypair.
 * @returns {{privateKey: string, publicKey: string}} PEM-encoded keys.
 */
function generateKeypair() {
  const { publicKey, privateKey } = crypto.generateKeyPairSync("ec", {
    namedCurve: "P-256",
    publicKeyEncoding: { type: "spki", format: "pem" },
    privateKeyEncoding: { type: "pkcs8", format: "pem" },
  });
  return { privateKey, publicKey };
}

/**
 * Create a token signed with an EC private key (algorithm: ES256).
 * @param {object} payload
 * @param {number} ttlSeconds
 * @param {string} privateKeyPem
 */
function issue(payload, ttlSeconds, privateKeyPem) {
  const now = Math.floor(Date.now() / 1000);
  const header = { alg: "ES256", typ: "QRT", v: 1 };
  const body = {
    iat: now,
    exp: now + ttlSeconds,
    jti: crypto.randomUUID(),
    data: payload,
  };
  const headerB64 = b64urlEncode(Buffer.from(JSON.stringify(header)));
  const bodyB64 = b64urlEncode(Buffer.from(JSON.stringify(body)));
  const signingInput = `${headerB64}.${bodyB64}`;
  // Default DER signature encoding matches Python's `cryptography` library default,
  // which is what makes tokens interoperable across both packages.
  const signature = crypto.sign("sha256", Buffer.from(signingInput), privateKeyPem);
  const signatureB64 = b64urlEncode(signature);
  return `${headerB64}.${bodyB64}.${signatureB64}`;
}

/**
 * Verify a token created by issue() using the corresponding EC public key.
 * @param {string} token
 * @param {string} publicKeyPem
 * @param {(jti: string) => (boolean|Promise<boolean>)} [isRevoked]
 */
async function verify(token, publicKeyPem, isRevoked) {
  const parts = typeof token === "string" ? token.split(".") : [];
  if (parts.length !== 3) {
    throw new InvalidSignatureError("Malformed token: expected 3 dot-separated parts");
  }
  const [headerB64, bodyB64, signatureB64] = parts;
  const signingInput = `${headerB64}.${bodyB64}`;
  const signature = b64urlDecode(signatureB64);

  const valid = crypto.verify(
    "sha256",
    Buffer.from(signingInput),
    publicKeyPem,
    signature
  );
  if (!valid) {
    throw new InvalidSignatureError("Signature mismatch");
  }

  assertSupportedVersion(headerB64);

  const body = JSON.parse(b64urlDecode(bodyB64).toString("utf-8"));
  const now = Math.floor(Date.now() / 1000);
  if (now > body.exp) {
    throw new TokenExpiredError(`Token expired at ${body.exp}`);
  }
  if (isRevoked) {
    const revoked = await isRevoked(body.jti);
    if (revoked) {
      throw new TokenRevokedError(`Token ${body.jti} has been revoked`);
    }
  }
  return body;
}

/**
 * Verify against multiple EC public keys (key-rotation window).
 * @param {string} token
 * @param {string[]} publicKeyPems - current public key first, then previous keys still in window
 * @param {(jti: string) => (boolean|Promise<boolean>)} [isRevoked]
 */
async function verifyWithPublicKeys(token, publicKeyPems, isRevoked) {
  if (!Array.isArray(publicKeyPems) || publicKeyPems.length === 0) {
    throw new InvalidSignatureError("At least one public key is required");
  }
  let lastSigError = null;
  for (const publicKeyPem of publicKeyPems) {
    try {
      return await verify(token, publicKeyPem, isRevoked);
    } catch (err) {
      if (err instanceof InvalidSignatureError) {
        lastSigError = err;
        continue;
      }
      throw err;
    }
  }
  throw lastSigError || new InvalidSignatureError("Signature mismatch");
}

module.exports = { generateKeypair, issue, verify, verifyWithPublicKeys };

File: python/admitiq/ec.py
"""
admitiq.ec — optional asymmetric (ES256 / ECDSA P-256) signing support.

Use this instead of core.issue/verify when the verifier (e.g. a scanner app
running on someone's phone, or a public endpoint) should NOT hold the same
secret used to issue tokens. The issuer signs with a private key; anyone
holding only the public key can verify a token but cannot forge new ones.

Requires the 'cryptography' package:
    pip install admitiq[ec]
"""
import json
import time
import uuid
from typing import Any, Callable, Dict, Optional, Sequence, Tuple

from .core import (
    InvalidSignatureError,
    TokenExpiredError,
    TokenRevokedError,
    _assert_supported_version,
    _b64url_decode,
    _b64url_encode,
)


def generate_keypair() -> Tuple[str, str]:
    """
    Generate a new EC (P-256 / secp256r1) keypair.

    Returns:
        (private_key_pem, public_key_pem) — both as PEM-encoded strings.
        Keep the private key secret (it issues tokens); the public key can be
        shared freely with anything that only needs to verify tokens.
    """
    try:
        from cryptography.hazmat.primitives import serialization
        from cryptography.hazmat.primitives.asymmetric import ec
    except ImportError as e:
        raise ImportError(
            "Asymmetric signing requires the 'cryptography' package. "
            "Install it with: pip install admitiq[ec]"
        ) from e

    private_key = ec.generate_private_key(ec.SECP256R1())
    public_key = private_key.public_key()

    private_pem = private_key.private_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PrivateFormat.PKCS8,
        encryption_algorithm=serialization.NoEncryption(),
    ).decode("ascii")
    public_pem = public_key.public_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PublicFormat.SubjectPublicKeyInfo,
    ).decode("ascii")
    return private_pem, public_pem


def issue(payload: Dict[str, Any], ttl_seconds: int, private_key_pem: str) -> str:
    """Create a token signed with an EC private key (algorithm: ES256)."""
    try:
        from cryptography.hazmat.primitives import hashes, serialization
        from cryptography.hazmat.primitives.asymmetric import ec
    except ImportError as e:
        raise ImportError(
            "Asymmetric signing requires the 'cryptography' package. "
            "Install it with: pip install admitiq[ec]"
        ) from e

    private_key = serialization.load_pem_private_key(
        private_key_pem.encode("ascii"), password=None
    )

    now = int(time.time())
    header = {"alg": "ES256", "typ": "QRT", "v": 1}
    body = {
        "iat": now,
        "exp": now + ttl_seconds,
        "jti": uuid.uuid4().hex,
        "data": payload,
    }
    header_b64 = _b64url_encode(json.dumps(header, separators=(",", ":")).encode())
    body_b64 = _b64url_encode(json.dumps(body, separators=(",", ":")).encode())
    signing_input = f"{header_b64}.{body_b64}".encode()

    signature = private_key.sign(signing_input, ec.ECDSA(hashes.SHA256()))
    signature_b64 = _b64url_encode(signature)
    return f"{header_b64}.{body_b64}.{signature_b64}"


def verify(
    token: str,
    public_key_pem: str,
    is_revoked: Optional[Callable[[str], bool]] = None,
) -> Dict[str, Any]:
    """Verify a token created by issue() using the corresponding EC public key."""
    try:
        from cryptography.exceptions import InvalidSignature
        from cryptography.hazmat.primitives import hashes, serialization
        from cryptography.hazmat.primitives.asymmetric import ec
    except ImportError as e:
        raise ImportError(
            "Asymmetric verification requires the 'cryptography' package. "
            "Install it with: pip install admitiq[ec]"
        ) from e

    try:
        header_b64, body_b64, signature_b64 = token.split(".")
    except ValueError:
        raise InvalidSignatureError("Malformed token: expected 3 dot-separated parts")

    public_key = serialization.load_pem_public_key(public_key_pem.encode("ascii"))
    signing_input = f"{header_b64}.{body_b64}".encode()
    signature = _b64url_decode(signature_b64)

    try:
        public_key.verify(signature, signing_input, ec.ECDSA(hashes.SHA256()))
    except InvalidSignature:
        raise InvalidSignatureError("Signature mismatch")

    _assert_supported_version(header_b64)

    body = json.loads(_b64url_decode(body_b64))

    if int(time.time()) > body["exp"]:
        raise TokenExpiredError(f"Token expired at {body['exp']}")

    if is_revoked is not None and is_revoked(body["jti"]):
        raise TokenRevokedError(f"Token {body['jti']} has been revoked")

    return body


def verify_with_public_keys(
    token: str,
    public_key_pems: Sequence[str],
    is_revoked: Optional[Callable[[str], bool]] = None,
) -> Dict[str, Any]:
    """
    Verify against multiple EC public keys (key-rotation window).

    Tries keys in order; succeeds on the first valid signature.
    """
    if not public_key_pems:
        raise InvalidSignatureError("At least one public key is required")

    last_sig_error: Optional[InvalidSignatureError] = None
    for public_key_pem in public_key_pems:
        try:
            return verify(token, public_key_pem=public_key_pem, is_revoked=is_revoked)
        except InvalidSignatureError as e:
            last_sig_error = e
            continue
    raise last_sig_error or InvalidSignatureError("Signature mismatch")
